[Fix][Core] Execute user requested actor exit in C++ side

MortalHappiness opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
When an asyncio task creates another asyncio task, raising `AsyncioActorExit` cannot make the caller exit because they are not the same task. Therefore, this PR makes `exit_actor` to request actor exit in core worker context, which will be checked regularly by core worker.

This PR also removes `AsyncioActorExit` exception because it is no longer needed.

## Related issue number

<!-- For example: "Closes #1234" -->
Closes: #49451

## Follow-up: Unify actor exit codepaths

### Core worker exit codepaths:
Either [ForceExit](https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L1239) or [Exit](https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L1160):
- Cancel task (called by `ray.cancel` python function): https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L4348
- Requested to kill actor by GCS:
  - https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L4510
  - https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1730
- Received `ExitRequest` RPC call: https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L4737
- Based on the status of `task_execution_callback` application-language callback:
  - https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L3351
  - https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L3424-L3443
- Based on the status of `check_signals` application-language callback:
  - https://github.com/ray-project/ray/blob/9008eb665bdcc978456211f01f2e93ebe052c6fa/src/ray/core_worker/core_worker.cc#L3159-L3170

### Actor exit codepaths (in GCS):
Either [DestroyActor](https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1024) or [KillActor](https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1745)
- Actor out of scope: https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L377
- Poll owner for actor ref deletd: https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1017
- Called by `ray.kill` (`CoreWorker::KillActor` -> `ActorInfoAccessor::AsyncKillActor` -> `KillActorViaGcs` RPC call -> `GcsActorManager::HandleKillActorViaGcs`): https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L694-L698
- Worker died: https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1217
- Node died: https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1294
- Actor scheduling failed: https://github.com/ray-project/ray/blob/67f9490c726ec528349e13e382ca86b6d54393f6/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L1517

### Examples:

- Normal actor exit (process completed): GCS will poll owner for actor ref deleted, and then notify core worker to delete actor.
- `exit_actor` is called (this PR): core worker exits based on application-language callback.
- `ray.kill` is called: `DestroyActor` in GCS will be called, and then it'll notify core worker to exit.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

edoakes commented:
@MortalHappiness let me know when this is ready for review

edoakes commented:
At first glance, I think we need to use a `threading.Event` here, not an `asyncio.Event`. This is because we need to communicate across threads, not just across asyncio tasks within the same thread.

MortalHappiness commented:
@edoakes What do you think would be a better approach? Implementing it purely on the Python side using `threading.Event`, or implementing it on the C++ side, as you mentioned in https://github.com/ray-project/ray/issues/49451#issuecomment-2590291813?

If implemented on the C++ side, I would export a function to Python and call this function when `exit_actor` is triggered. Inside this function, I would modify some state and periodically check whether this state indicates that an exit is required.

MortalHappiness commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -989,6 +989,45 @@ def verify():
+        async def normal_task(self):
+            pass
+
+    a = AsyncActor.remote()
+    ray.get(a.__ray_ready__.remote())
+    ray.get(a.start_exit_task.remote())
+    # Wait for the actor to exit.
+    time.sleep(1)
+    with pytest.raises(ray.exceptions.RayActorError):
+        ray.get(a.normal_task.remote())
```
Hi @edoakes I've already switched to implementing this in C++. However, I can't make it throw `RayActorError` immediately when the result of `start_exit_task` is obtained. Instead, I ensure that subsequent calls will result in `RayActorError`. Is that acceptable, or do I need to guarantee that `RayActorError` is thrown as soon as the result of `start_exit_task` is obtained?

MortalHappiness replied:
Changed the implementation according to your comment. Thanks!

edoakes commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -989,6 +989,45 @@ def verify():
+    class AsyncActor:
+        def __init__(self):
+            def f():
+                print("atexit handler")
+                with open(async_temp_file, "w") as f:
+                    f.write("Async Actor\n")
```
please use the `SignalActor` utility instead for cross-process synchronization in tests


lmk if you have questions on how to do it

MortalHappiness replied:
I copied this from other tests in this file. Should I also update the other tests (perhaps in follow-up PRs)?

edoakes replied:
yes please do; a follow-up PR sounds great

MortalHappiness replied:
After offline discussion, there is a weird interaction between `atexit` handler. So I don't use `SignalActor` in this PR in the end.

edoakes commented on python/ray/_raylet.pyx:
```diff
@@ -4621,6 +4621,10 @@ cdef class CoreWorker:
         return (CCoreWorkerProcess.GetCoreWorker().GetWorkerContext()
                 .CurrentActorIsAsync())
 
+    def exit_current_actor(self):
+        return (CCoreWorkerProcess.GetCoreWorker().GetWorkerContext()
+                .SetCurrentActorShouldExit())
```
```suggestion
    def set_current_actor_should_exit(self):
        return (CCoreWorkerProcess.GetCoreWorker().GetWorkerContext()
                .SetCurrentActorShouldExit())
```

No reason to have mismatching names (which is confusing to read), and this name better conveys the behavior. As written, `exit_current_actor` implies that the actor would exit immediately.

edoakes reviewed (commented):
@MortalHappiness great start!

Let's please unify as many of the actor exit codepaths as possible to share the same logic. For example, can we also unify the single threaded and multi threaded actor exit logic to use the same flag?

MortalHappiness commented on python/ray/actor.py:
```diff
@@ -1781,6 +1781,7 @@ def exit_actor():
         # quit the asycnio event loop thread, not the main thread. Instead, we
         # raise a custom error to the main thread to tell it to exit.
+        worker.core_worker.exit_current_actor()
         if worker.core_worker.current_actor_is_asyncio():
             raise AsyncioActorExit()
```
So should I remove `AsyncioActorExit` completely?

MortalHappiness replied:
Update: I've removed `AsyncioActorExit`

MortalHappiness commented:
> Let's please unify as many of the actor exit codepaths as possible to share the same logic. For example, can we also unify the single threaded and multi threaded actor exit logic to use the same flag?

Should I do this in this PR or in follow-up PRs?

edoakes commented:
> > Let's please unify as many of the actor exit codepaths as possible to share the same logic. For example, can we also unify the single threaded and multi threaded actor exit logic to use the same flag?
> 
> Should I do this in this PR or in follow-up PRs?

I'm ok to do it in follow-up PRs, but could you summarize for me all of the existing exit paths before this PR and what you see as the end goal state?

MortalHappiness commented:
Hi @edoakes I've tried to summarize exit codepaths in PR description. Let me know if I missed something.

edoakes commented on python/ray/_raylet.pyx:
```diff
@@ -4621,6 +4623,14 @@ cdef class CoreWorker:
+        return (CCoreWorkerProcess.GetCoreWorker().GetWorkerContext()
+                .SetCurrentActorShouldExit())
+
+    def current_actor_should_exit(self):
+        return (CCoreWorkerProcess.GetCoreWorker().GetWorkerContext()
+                .CurrentActorShouldExit())
```
nit: "get" and "set" usually go together
```suggestion
    def get_current_actor_should_exit(self):
        return (CCoreWorkerProcess.GetCoreWorker().GetWorkerContext()
                .GetCurrentActorShouldExit())
```

edoakes commented on python/ray/actor.py:
```diff
@@ -1768,7 +1766,9 @@ def exit_actor():
     API if you'd like to kill an actor using actor handle.
 
-    When the API is called, the actor raises an exception and exits.
+    When the API is called, the actor sets a flag to exit.
+    The worker periodically checks the flag and exits the actor.
+    So there may be some delay before the actor actually exits.
```
For non-asyncio actors, we should still be raising the exception immediately, right? Don't want a behavior change there

edoakes commented on python/ray/actor.py:
```diff
@@ -1777,19 +1777,10 @@ def exit_actor():
-        if worker.core_worker.current_actor_is_asyncio():
-            raise AsyncioActorExit()
-
-        # Set a flag to indicate this is an intentional actor exit. This
-        # reduces log verbosity.
-        raise_sys_exit_with_custom_error_message("exit_actor() is called.")
```
We can probably leave this part in, else it's a behavior change (code after `exit_actor()` will now be called)

edoakes commented on python/ray/exceptions.py:
```diff
@@ -777,13 +777,6 @@ class PlasmaObjectNotAvailable(RayError):
     pass
 
 
-@PublicAPI
-class AsyncioActorExit(RayError):
```
was this ever actually raised to users before? I don't think it was

edoakes commented:
Looks good now -- my only concern is that we don't have any behavior change for the sync worker case. Let's make sure that still raises and fails the current task w/ an `ActorError`

MortalHappiness commented:
Hi @edoakes I've addressed all comments. Ready for another review. Thanks.

edoakes commented on python/ray/actor.py:
```diff
@@ -1768,7 +1767,11 @@ def exit_actor():
     API if you'd like to kill an actor using actor handle.
 
-    When the API is called, the actor raises an exception and exits.
+    When the API is called, if the actor it not async actor,
+    it raises an exception and exits. For async actor,
+    it sets a flag to exit and the worker periodically checks
+    the flag and exits the actor. So there may be some delay
+    before the actor actually exits.
```
```suggestion
    When this API is called, an exception is raised and the actor
    will exit immediately. For asyncio actors, there may be a short
    delay before the actor exits if the API is called from a background
    task.
```

edoakes commented on python/ray/actor.py:
```diff
@@ -1777,19 +1780,12 @@ def exit_actor():
             f"{worker.mode}. Call this API inside an actor methods"
             "if you'd like to exit the actor gracefully."
         )
+    if not worker.core_worker.current_actor_is_asyncio():
+        raise_sys_exit_with_custom_error_message("exit_actor() is called.")
+    worker.core_worker.set_current_actor_should_exit()
```
A little confused at why this is conditional because we should try to make the behavior as uniform as possible:

- For non-async actors, don't we still want to set the field in the core worker? Otherwise, if users accidentally or intentionally catch the `SystemExit` exception that is raised, the actor won't actually be exited.
- For async actors, don't we still want to raise the `SystemExit` exception immediately?

edoakes commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -918,6 +918,10 @@ def f():
     class Actor:
         def exit(self):
             exit_actor()
+            raise AssertionError(
+                "This line should not be reached because non-async actor"
+                "should exit immediately after exit_actor is called."
+            )
```
I don't think this assertion would actually catch a regression because this line might execute and then the `exit` method would still return a `RayActorError`. For any condition like this, you should verify that if you intentionally break the behavior (in this case, make it so that sync actors *don't* raise immediately), the test fails. I don't think it will in this case.

Instead, what you will need to do is add an "out of bounds" check -- so for example catch the `SystemExit` exception and then send a signal on an actor or write a tempfile like the other test cases. Then after the actor exits, assert that the exception was raised and therefore the out of band behavior happened.

edoakes commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -989,6 +998,45 @@ def verify():
+            exit_actor()
+
+    a = AsyncActor.remote()
+    ray.get(a.__ray_ready__.remote())
+    with pytest.raises(ray.exceptions.RayActorError) as exc_info:
+        ray.get(a.start_exit_task.remote())
```
hm this test doesn't look correct. the method starts a background task so there is no guarantee that runs before the method finishes successfully. I think what you need to do is start the background task and have it block on a `SignalActor` before it actually calls `exit_actor()`, check that `start_exit_task` completes successfully, then unblock the background task so it calls `exit_actor()`, then use `wait_for_condition` on `ray.get(a.noop_method.remote())` to check that eventually subsequent methods will return the actor error

MortalHappiness commented:
@edoakes I've addressed all of your comments, but the actual implementation differs slightly. Let me describe the changes here:

- I added back the `AsyncioActorExit` exception. Since you mentioned that a normal async task should raise immediately, I can't just call `set_current_actor_should_exit` because this flag is checked periodically rather than triggering an immediate exception. I also can't raise `SystemExit`, as noted in the original comment https://github.com/ray-project/ray/blob/e151dc210348ab2ce37562583859f1afea035198/python/ray/actor.py#L1784-L1786

- I split the test into **five separate tests**:

  1. **`test_exit_actor_normal_actor_raise_immediately`**:  
     - This is the original test for a normal actor, but I added one line to ensure that any code after `exit_actor` is **not executed**.
  2. **`test_exit_actor_normal_actor_user_catch_err_should_still_exit`**:  
     - This is an "out-of-bound" test for a normal actor.  
     - In this case, the user catches `SystemExit`, so the program **does not raise immediately**.  
     - However, the actor will still exit shortly after the flag is checked by the core worker, similar to the exit path for an async background task.
  3. **`test_exit_actor_async_actor`**:  
     - This is the original test for an async actor, but I added one line to verify that any code after `exit_actor` is **not executed**.  
     - Why does this test still check both `"exit_actor()" in str(exc_info.value)"` and `"User requested to exit the actor" in str(exc_info.value)"`?  
       - Most of the time, the first case occurs.  
       - However, since we call `set_current_actor_should_exit` (which is checked periodically by the core worker), there is a small chance that the second case occurs instead.
  4. **`test_exit_actor_async_actor_user_catch_err_should_still_exit`**:  
     - This is an "out-of-bound" test for an async actor.  
     - In this case, the user catches `AsyncioActorExit`, so the program **does not raise immediately**.  
     - However, the actor will still exit shortly after the flag is checked by the core worker, similar to the exit path for an async background task.
  5. **`test_exit_actor_async_actor_nested_task`**:  
     - I updated this test to use `SignalActor`.  
     - No error is raised in this case, but the actor **still exits** short after when the flag is checked by the core worker.
     - Therefore the `atexit` handler is triggered.

MortalHappiness commented:
> @MortalHappiness why can't we set the `should_exit` flag and then raise the normal `SystemExit`? AFAICT this makes the behavior as consistent as possible:
> 
> * The currently-running task will raise and exit immediately (same as a sync actor method).
> * The actor will exit soon after once the `should_exit` flag is checked.
> 
> Am I missing something?

After some debug printing and some experiment. I found that `asyncio.run_coroutine_thread_safe` handles `SystemExit` differently.

The following is a toy example.

```python
import time
import asyncio
import threading

async def f():
    print("Inside coroutine f()")
    # raise ValueError
    # raise SystemExit
    raise BaseException

def start_loop(loop):
    asyncio.set_event_loop(loop)
    loop.run_forever()

def main():
    loop = asyncio.new_event_loop()
    thread = threading.Thread(target=start_loop, args=(loop,))
    thread.start()

    future = asyncio.run_coroutine_threadsafe(f(), loop)

    try:
        time.sleep(2)
        print("loop alive?", loop.is_running())
        future.result()
    except BaseException as e:
        print("Caught exception:", repr(e))

    loop.call_soon_threadsafe(loop.stop)
    thread.join()

if __name__ == "__main__":
    main()
```

![image](https://github.com/user-attachments/assets/f1842f14-0af4-426f-b6f5-efc7161e516a)

Explanation:
- raise `BaseException` in `f`: Loop alive, and exception was caught.
- raise `ValueError` in `f`: Loop alive, and exception was caught.
- raise `SystemExit` in `f`: Loop dead.

And in Ray, we use `run_coroutine_thread_safe` to run async tasks:

https://github.com/ray-project/ray/blob/b84558144a3a90058bc46af1f27fa6625edbf65c/python/ray/_raylet.pyx#L4587

This is why we cannot raise `SystemExit` in `exit_actor`.

edoakes commented:
@MortalHappiness I see, thanks for the explanation. Is it possible to add a top-level exception handle for `SystemExit` in the `async_func()` that will call `future.set_exception()`?

MortalHappiness commented:
> @MortalHappiness I see, thanks for the explanation. Is it possible to add a top-level exception handler for `SystemExit` in the `async_func()` that will call `future.set_exception()`?

@edoakes Not easy to do that because the future object is returned by `asyncio.run_coroutine_threadsafe`, so we cannot call `future.set_exception` inside `async_func`.

edoakes commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -910,83 +911,175 @@ def f():
+    except Exception:
+        pass
+
+    # with pytest.raises(ray.exceptions.RayActorError) as exc_info:
+    #     ray.get(a.exit.remote())
+    # assert (
+    #     # Exited when task execution returns
+    #     "exit_actor()" in str(exc_info.value)
+    #     # Exited during periodical check in worker
+    #     or "User requested to exit the actor" in str(exc_info.value)
+    # )
```
is this intended to be deleted or uncommented?

MortalHappiness replied:
fixed

edoakes commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -910,83 +911,175 @@ def f():
+    def verify():
+        return temp_file_atexit.exists()
+
+    # wait_for_condition(verify)
+    # time.sleep(3)
+    # assert not temp_file_after_exit_actor.exists()
```
ditto

MortalHappiness replied:
fixed

edoakes commented on src/ray/core_worker/context.h:
```diff
@@ -117,6 +117,10 @@ class WorkerContext {
 
   bool CurrentActorIsAsync() const ABSL_LOCKS_EXCLUDED(mutex_);
 
+  void SetCurrentActorShouldExit() ABSL_LOCKS_EXCLUDED(mutex_);
+
+  bool GetCurrentActorShouldExit() const ABSL_LOCKS_EXCLUDED(mutex_);
```
Please add header comments for these indicating what they're used for

MortalHappiness replied:
fixed

MortalHappiness replied:
Added 3 tests: `test_exit_actor_normal_actor_in_constructor_should_exit`, `test_exit_actor_async_actor_in_constructor_should_exit`, and `test_exit_actor_async_actor_nested_task_in_constructor_should_exit`

edoakes commented on src/ray/core_worker/context.cc:
```diff
@@ -394,6 +395,17 @@ bool WorkerContext::CurrentActorIsAsync() const {
 
+void WorkerContext::SetCurrentActorShouldExit() {
+  absl::WriterMutexLock lock(&mutex_);
+  RAY_CHECK(!current_actor_id_.IsNil());
```
```suggestion
  RAY_CHECK(!current_actor_id_.IsNil()) << `SetCurrentActorShouldExit should only be used inside actors`;
```

edoakes commented on python/ray/tests/test_actor_failures.py:
```diff
@@ -910,83 +911,175 @@ def f():
-    Verify the basic case.
-    """
+
+def test_exit_actor_normal_actor_raise_immediately(shutdown_only, tmp_path):
```
Can you add a test case for calling `ray.exit_actor()` within the constructor if there isn't one already?

MortalHappiness replied:
Added 3 tests: `test_exit_actor_normal_actor_in_constructor_should_exit`, `test_exit_actor_async_actor_in_constructor_should_exit`, and `test_exit_actor_async_actor_nested_task_in_constructor_should_exit`

edoakes reviewed (commented):
Code changes LGTM, just a few nits

Your latest suggestion sounds better than what is in the code, but given that we already have the existing behavior with `AsyncioActorExit` and it would be a very minor API breakage to change, better to leave it as is.

edoakes reviewed (approved):
🚢

MortalHappiness commented:
@edoakes I found that you enabled auto-merged. But I found that I accidentally changed 1 line that should not be changed so I pushed it again and the auto-merge is disabled. Could you enable it again? Thanks.
